### PR TITLE
8325168: JShell should support Markdown comments

### PR DESCRIPTION
**Pull request by lahodaj**
This is an attempt to add Markdown support in documentation comments to JShell.

It works by converting the Markdown text to HTML during the process of resolving `{@inheritDoc}` tags.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325168](https://bugs.openjdk.org/browse/JDK-8325168): JShell should support Markdown comments (**Enhancement** - P3)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [d85a2b8a](https://git.openjdk.org/jdk/pull/17686/files/d85a2b8a0e01e362d63c10eb6345b5cb43f306c5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17686/head:pull/17686` \
`$ git checkout pull/17686`

Update a local copy of the PR: \
`$ git checkout pull/17686` \
`$ git pull https://git.openjdk.org/jdk.git pull/17686/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17686`

View PR using the GUI difftool: \
`$ git pr show -t 17686`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17686.diff">https://git.openjdk.org/jdk/pull/17686.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17686#issuecomment-2018817650)

### CODE REVIEW

**Comment by jonathan-gibbons** [comment]
Compared to the previous PR, which was reasonably simple and quite elegant,
this one seems unduly complicated and somewhat invasive. It is also using
the transformer mechanism in a way that is not as intended. 

The `jdk.internal.md` module is intended to be for JDK support for Markdown.
As such, it has an imported copy of `commonmark-java` and now the new transformer
support, which is just intended to be used to support any JDK extensions to Markdown,
such as the new support for reference links for program elements. (There are 
suggestions for other future features in the JEP.) As such, I would 
expect all JDK clients to use the existing "standard" transformer, and to not 
substitute their own.  Ideally, the "standard" transformer would be used
by default, but I wanted to avoid an unconditional dependency from `jdk.compiler`
to `jdk.internal.md` -- so hence the use of the service provider mechanism.
(I guess I could update `jdk.compiler` to use `requires static` to establish
a compile-time dependency on the `jdk.internal.md` module, and then try
and install the standard transformer, which would always be available if we
put a hard dependency in modules like `jdk.javadoc` and `jdk.jshell`.)

Your previous PR was pretty good (although I accept we did not get as far as
a full review and approval.)  The recent changes the overall Markdown work
were made with that PR in mind.  As such, the intention was that only two
fairly small changes would be required.

1. Change the means to detect Markdown comments, from looking for `/**md`
   to using the new `CommentKind` support in `DocTrees`.
2. Install the 'standard' transformer, to enable the JDK extensions to
   Markdown

With those two changes, you can use the previous technique to parse and 
render a Markdown doc comment as before.

**Review by jonathan-gibbons — CHANGES_REQUESTED** [review]

**Comment by lahodaj** [comment]
Thanks for the comments! I've dropped the `ToHTML` transformer, and made the transformations easier. Sadly, the standard transformer seems to break raw text tree positions, so it is not currently possible to properly replace the Markdown text with the Javadoc text when the transformer is installed. Please see:
https://github.com/openjdk/jdk/pull/16388#discussion_r1479306878
for more details. Once that is solved, it should be possible to merge the newest changes in the Markdown branch.

File: src/jdk.jshell/share/classes/jdk/internal/shellsupport/doc/JavadocHelper.java [hunk 1]
```diff
@@ -478,6 +512,21 @@ public Void visitReturn(ReturnTree node, Void p) {
                     }
                     return super.visitInheritDoc(node, p);
                 }
+                @Override
+                public Void visitLink(LinkTree node, Void p) {
+                    if (sp.isRewrittenTree(null, dcTree, node)) {
+                        //this links is a synthetic rewritten link, replace
```
Comment (jonathan-gibbons) [inline src/jdk.jshell/share/classes/jdk/internal/shellsupport/doc/JavadocHelper.java:518]: spelling `this link`

**Review by jonathan-gibbons — APPROVED** [review]
I'll approve it for what it is.

There are some minor comments about imports and tests; these can either be fixed here in this PR, or in cleanup tasks before RDP2.

It is disappointing but not surprising to see duplication of somewhat similar code in the `jdk.javadoc` module, especially the handling of Markdown code. This just adds to the duplication of effort interpreting other aspects of doc comments. I understand how we got to this point, but down the road, we should look to share more code, translating doc comments into some intermediate representation (such as the HTML classes currently in  `jdk.javadoc.internal.doclets.formats.html.markup`) such that we only need minimal subsequent conversion to a near-text form., like ANSI-styled text. That would be a good layer of abstraction to have in the `jdk.javadoc` module anyway, and if we can expose it (internally?) to other JDK modules, to support their use and display of doc comments, so much the better: we all win.

**Pull request closed without merging**